Clarify that Reserved may also represent private-use in data model (#444)
unicode-org/message-format-wg: spec/data-model.md

}
```

A `Reserved` represents an _expression_ with a _reserved_ or _private-use_ _annotation_.

> The `sigil` corresponds to the starting sigil of the _annotation_.

Reviewer: I think this is incorrect.

Reserved represents a portion of the syntax that is wholly invalid--not permitted to be used--but which might become part of a future incarnation. If a part of reserved were used in a future incarnation, existing (pre the unreserving) implementations would continue with whatever behavior reserved provides. Probably this means, as you have it here, passing reserved gunk through the data model without processing.

Private use is different. Private use is a portion of the syntax that _is valid_, but which may not be functional in a given implementation. 

What I would do is define private use and reserved separately and use the same language where appropriate.

Also: note that private-use is a feature of our specification, which is why I put it before reserved in the text.

Author: Could you give an example of a data model use case where it would be useful for the _private-use_ and _reserved_ syntax rules to be represented by different data structures, rather than this single `Reserved`?

Given that supported private-use syntax would end up using a different interface altogether, I'm failing to come up with a realistic scenario where I'd like unsupported private-use and reserved to be handled differently. I suppose a linter error for a message including one or the other could use a slightly different text, but that can be detected from the `sigil`.

The `source` is the "raw" value (i.e. escape sequences are not processed)

> and does not include the starting `sigil`.

Reviewer: Is this true? The escape sequences prevent falsely finding an ending marker. The unescaped `source` is the "raw" value. Reserializing the `source` would reinsert the escapes (this is how one avoids double-escaping traps)

Author: The `sigil` in this case is the `^`, `*` or other identified sigil from which we can tell that this annotation is private-use or reserved. Including it or not in `source` is a matter of choice; I'm here proposing that it should not be, as that's how all the other sigils are handled in the spec.

Reviewer: My comment here isn't about sigils. It is that preserving the "raw" value without unescaping produces a risk of double-escaping (the receiver cannot tell if the escape sequence is intentional or not). 

What is the value of preserving the "raw" (escaped) representation of the annotation in the data model, given that the purpose of the escapes is either ambiguity in parsing MF2 or in the runtime environment?

Author: The problem is that if we process the string, then we need to process it into a sequence of reserved-char sequences and quoted literals, because their escapes are different.

Say we didn't do that, and just applied the escapes. Then we could end up with a cooked source value like
```
one|two|three|four
```

With that, we can't know which of the following would be its original source form:
1. `one|two|three\|four`
2. `one|two\|three|four`
3. `one\|two|three|four`
4. `one\|two\|three\|four`

As the implementation or future spec version undoubtedly cares very much whether a `|` is escaped or not, we would not be able to reconstruct the annotation source.

My presumption is that almost always when encountering unsupported syntax we either don't care about its syntax source, or we just want to pass it through untouched; hence the current approach. If you think this behaviour should be reconsidered, could you raise a separate issue or submit a PR rather than increasing the scope of this one?

Reviewer: I think this describes a bug in the syntax, where we were "clever" to smuggle in the quoted production and the whitespace production. I think the argument was something like: "if we unreserved a sigil, then it would just parse correctly". But I think we're better off if reserved and private are opaque to the ABNF. An unreserved sigil is subject to whatever ABNF is applied to it (either the existing annotation syntax or new syntax). Pre-unreserved implementations still won't parse into the sigil's space, so won't be affected.

In other words, I think we should modify the ABNF thusly:

```abnf
private-use    = private-start text
private-start  = "^" / "&"

; reserve additional sigils for use by 
; future versions of this specification
reserved       = reserved-start text
reserved-start = "!" / "@" / "#" / "%" / "*" / "<" / ">" / "/" / "?" / "~"
```

The only escapes in text are around `{` and `}` (and `\` in case one needs `\}` as a character sequence). Brackets retain syntactic meaning everywhere. The current `reserved-body` approach means that implementations parse `reserved` and `private-use` into word tokens and literals, even though the implementation is not allowed to interpret them. I think true opacity is the right approach. (In that case (1), (2), (3), and (4) are different character sequences).

Reviewer: (and, yeah, we've had this discussion before [in #374](https://github.com/unicode-org/message-format-wg/pull/374#discussion_r1167447318))

Your argument there was:

> Currently this is a valid expression:
>```
>{:foo key=|{bar}|}
>```
>If we were to unreserve @ and try to give it the same semantics as with :, this would be a parse error due to the unescaped >inner {:
>```
>{@foo key=|{bar}|}
>```

The problem then is: `quoted` allows `{` and `}` unescaped (because they are "quoted"). Adding these characters to the escaped list for `quoted` simplifies the ABNF a tiny amount (quoted-escape and reserve-escape are the same and we lose some productions) and lets implementations parse out expressions by matching (unescaped) `{`/`}` pairs. The cost is that `{`/`}` must be escaped in a literal. 

While fewer escapes is better than more escapes in quoted literals, I don't think that is very onerous compared to having `private-use` (`reserved` scares me less, as we might never use it) be "semi-parsed". I was convinced before that the ABNF would be okay using what we have now because the character sequences could be squeezed in (that I wouldn't actually have to parse the contents). But the data model shows why I was shy in the first place: `private-use` and `reserved` turn out to have parsed structure where we want opacity.

Reviewer: Opened #446 to discuss.

> and does not include the starting `sigil`.

Reviewer: Also, by removing the starting sigil, you remove the ability for an implementation to tell how reserved (or private) sequence was introduced. If the data model were serialized and sent to an implementation that supported one or another sigil, the receiver would have no way of knowing what the sigil was. Is there a reason you removed the sigil?

Author: It's because it's already in the separate `sigil` field.

> Implementations MUST NOT rely on the set of `sigil` values remaining constant,
> as future versions of this specification MAY assign other meanings to such sigils.

Reviewer: I think this normative text is unenforceable. If I hardcoded the existing sigils for functions, pua, and reserved into my code, that's my business. I think any test we make of an MFv2 "1.0" implementation would absolutely depend on implementations treating the set of sigil values as constant (for that version).

I think what you mean is:

```suggestion
> **Note**
> Implementers should be aware that future versions of this specification
> might assign meaning to _reserved_ `sigil` values.
```

PS> I use a non-normative "should" because Notes are non-normative.

Author: Sure; the MUST NOT is indeed probably pretty impossible to break. I think the note might read better with this expansion?

```suggestion
> **Note**
> Implementers should be aware that future versions of this specification
> might assign meaning to _reserved_ `sigil` values
> and add new interfaces to this data model.
```

Reviewer: Yes, that.

If the _expression_ includes a _literal_ or _variable_ before the _annotation_,
it is included as the `operand`.

When parsing the syntax of a _message_ that includes a _private-use_ _annotation_
supported by the implementation,

> the implemenation MAY represent it in the data model using a different interface

Reviewer: I think we might need some stronger guidance here. I agree with the "MAY" you have, but would suggest something along the lines of:

> _Private-use_ _annotations_ are specific to a particular "private agreement"
> between the various implementations that support a given form of _private-use_.
>
> When parsing the syntax of a _message_ that includes a _private-use_ _annotation_
> unrecognized by the implementation, the _annotation_ MUST be processed
> identically to a _reserved_ _annotation_.
>
> An implementation that supports a given _private-use_ _annotation_ MUST
> define the specific interface to support the semantics, structure, and meaning
> that it provides. Use of existing data model interfaces is RECOMMENDED, although
> an implementation MAY use any interface appropriate for its needs.

Author: I don't think the "parsing the syntax" sentence belongs in the data model spec, but in the syntax. It's defining a MUST about what the syntax means, and if included here that would get watered down by the qualifier at the top of this doc:
> Implementations are not required to use this data model for their internal representation of messages.

On review, I now also note that the Extensions section will need to be updated to account for private-use. Something like your third paragraph is probably needed there.
